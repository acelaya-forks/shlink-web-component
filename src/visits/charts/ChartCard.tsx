import { Card } from '@shlinkio/shlink-frontend-kit';
import type { FC, PropsWithChildren, ReactNode } from 'react';

type ChartCardProps = PropsWithChildren<{
  title: ReactNode;
  footer?: ReactNode;
}>;

export const ChartCard: FC<ChartCardProps> = ({ title, footer, children }) => (
  <Card role="document">
    <Card.Header>{title}</Card.Header>
    <Card.Body>{children}</Card.Body>
    {footer && <Card.Footer className="sticky bottom-0">{footer}</Card.Footer>}
  </Card>
);
