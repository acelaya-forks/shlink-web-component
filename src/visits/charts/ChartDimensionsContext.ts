import { createContext, Fragment , useContext, useMemo } from 'react';
import { ResponsiveContainer } from 'recharts';

export type Dimensions = {
  width: number;
  height: number;
};

export type WrapperDimensions = {
  width?: '100%';
  height?: number;
};

export type ChartWrapperType = typeof ResponsiveContainer | typeof Fragment;

const ChartDimensionsContext = createContext<Dimensions | null>(null);

export const { Provider: ChartDimensionsProvider } = ChartDimensionsContext;

export type ChartDimensions = {
  ChartWrapper: ChartWrapperType;
  dimensions?: Dimensions;
  wrapperDimensions: WrapperDimensions;
};

export const useChartDimensions = (defaultHeight: number): ChartDimensions => {
  const dimensions = useContext(ChartDimensionsContext) ?? undefined;
  const wrapperDimensions = useMemo(() => dimensions ? {} : {
    width: '100%' as const,
    height: defaultHeight,
  }, [defaultHeight, dimensions]);
  const ChartWrapper = dimensions ? Fragment : ResponsiveContainer;

  return { ChartWrapper, dimensions, wrapperDimensions };
};
