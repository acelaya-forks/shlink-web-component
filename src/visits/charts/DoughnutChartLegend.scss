@use '../../../node_modules/@shlinkio/shlink-frontend-kit/dist/base';

.doughnut-chart-legend {
  list-style-type: none;
  padding: 0;
  margin: 0;

  @media (max-width: base.$smMax) {
    margin-top: 1rem;
  }
}

.doughnut-chart-legend__item:not(:first-child) {
  margin-top: .3rem;
}

.doughnut-chart-legend__item-color {
  width: 20px;
  min-width: 20px;
  height: 20px;
  margin-right: 5px;
  border-radius: 10px;
}

.doughnut-chart-legend__item-text {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}
