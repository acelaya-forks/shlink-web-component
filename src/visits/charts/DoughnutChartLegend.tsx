import { formatNumber } from '@shlinkio/shlink-frontend-kit/tailwind';
import type { FC } from 'react';
import { ColorBullet } from '../../utils/components/ColorBullet';
import type { DoughnutChartEntry } from './DoughnutChart';

export type DoughnutChartLegendProps = {
  chartData: DoughnutChartEntry[];
  showNumbers: boolean;
};

export const DoughnutChartLegend: FC<DoughnutChartLegendProps> = ({ chartData, showNumbers }) => (
  <ul className="tw:p-0 tw:m-0 tw:space-y-1">
    {chartData.map(({ name, color, value }, index) => (
      <li key={name} className="tw:flex tw:items-center">
        <ColorBullet color={color} testId={`color-bullet-${index}`} />
        <small className="tw:truncate tw:grow" data-testid={`name-${index}`}>
          {name}
          {showNumbers && <b> ({formatNumber(value)})</b>}
        </small>
      </li>
    ))}
  </ul>
);
