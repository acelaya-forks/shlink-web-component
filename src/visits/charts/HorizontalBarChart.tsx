import {
  formatNumber,  HIGHLIGHTED_COLOR,
  HIGHLIGHTED_COLOR_ALPHA,
  isDarkThemeEnabled,
  MAIN_COLOR,
  MAIN_COLOR_ALPHA } from '@shlinkio/shlink-frontend-kit';
import type { FC } from 'react';
import { useMemo } from 'react';
import { Bar, CartesianGrid, Cell, ComposedChart, Tooltip, XAxis, YAxis } from 'recharts';
import type { Stats } from '../types';
import { useChartDimensions } from './ChartDimensionsContext';
import { CHART_TOOLTIP_COMMON_PROPS, prevColor, prevColorAlpha } from './constants';

export type HorizontalBarChartProps = {
  stats: Stats;
  prevStats?: Stats;
  highlightedStats?: Stats;
  highlightedLabel?: string;
  max?: number;
  onClick?: (label: string) => void;
};

type HorizontalBarChartEntry = {
  name: string;
  amount: number;
  nonHighlightedAmount: number | null;
  highlightedAmount: number | null;
  prevAmount: number | null;
};

const isHiddenLabel = (label: string) => label.startsWith('hidden_');

export const HorizontalBarChart: FC<HorizontalBarChartProps> = (
  { stats, prevStats, highlightedStats, highlightedLabel, max, onClick },
) => {
  const chartData = useMemo((): HorizontalBarChartEntry[] => Object.entries(stats).map(([name, amount]) => {
    const highlightedAmount = highlightedStats?.[name] ?? 0;
    const prevAmount = prevStats?.[name] ?? 0;
    const isHidden = isHiddenLabel(name);

    return {
      name,
      amount: Math.max(amount, prevAmount),
      // Setting value `null` on "hidden" elements allows for them to be excluded from tooltips
      nonHighlightedAmount: isHidden ? null : amount - highlightedAmount,
      highlightedAmount: isHidden ? null : highlightedAmount,
      prevAmount: isHidden ? null : prevAmount,
    };
  }), [stats, prevStats, highlightedStats]);
  const verticalAxisWidth = useMemo(() => {
    const longestNameLength = chartData.reduce((prev, { name }) => (prev > name.length ? prev : name.length), 0);
    // Set a size of around 7 times the longest text, unless it exceeds a maximum of 150
    return Math.min(150, longestNameLength * 7);
  }, [chartData]);

  const { ChartWrapper, dimensions, wrapperDimensions } = useChartDimensions(
    Math.max(300, chartData.length * (prevStats ? 44 : 22)),
  );

  return (
    <ChartWrapper {...wrapperDimensions}>
      {/* Using a ComposedChart instead of a PieChart because they have a more subtle hover effect */}
      <ComposedChart layout="vertical" data={chartData} barCategoryGap={3} {...dimensions}>
        <XAxis
          type="number"
          dataKey="amount"
          tickFormatter={formatNumber}
          domain={max ? [0, max] : undefined}
        />
        <YAxis
          type="category"
          dataKey="name"
          width={verticalAxisWidth}
          interval={0}
          style={{ fontSize: '.8rem' /* TODO: Should show text overflow ellipsis */ }}
          tickFormatter={(label) => (isHiddenLabel(label) ? '' : label)}
        />
        <CartesianGrid strokeOpacity={isDarkThemeEnabled() ? 0.05 : 0.9} />
        <Tooltip
          filterNull // This will prevent "hidden" values to render a tooltip
          {...CHART_TOOLTIP_COMMON_PROPS}
          formatter={(value: number, name: keyof HorizontalBarChartEntry) => {
            const prettifiedValue = formatNumber(value);
            const label = (() => {
              if (name === 'highlightedAmount') {
                return highlightedLabel;
              }
              if (name === 'prevAmount') {
                return 'Previous period';
              }

              return highlightedStats ? 'Non-selected' : 'Visits';
            })();

            return [prettifiedValue, label];
          }}
        />
        <Bar
          dataKey="nonHighlightedAmount"
          stackId="main"
          cursor="pointer"
          fill={MAIN_COLOR /* This needs to be set as it is the color used in the tooltip */}
          onClick={({ name }: HorizontalBarChartEntry) => onClick?.(name)}
        >
          {chartData.map((entry) => (
            // Using a Cell, to define a different fill color, without affecting the one used for the tooltip
            <Cell key={entry.name} fill={MAIN_COLOR_ALPHA} stroke={MAIN_COLOR} strokeWidth={2} />
          ))}
        </Bar>
        {highlightedStats && (
          <Bar
            dataKey="highlightedAmount"
            stackId="main"
            cursor="pointer"
            fill={HIGHLIGHTED_COLOR /* This needs to be set as it is the color used in the tooltip */}
            onClick={({ name }: HorizontalBarChartEntry) => onClick?.(name)}
          >
            {chartData.map((entry) => (
              <Cell key={entry.name} fill={HIGHLIGHTED_COLOR_ALPHA} stroke={HIGHLIGHTED_COLOR} strokeWidth={2} />
            ))}
          </Bar>
        )}
        {prevStats && (
          <Bar dataKey="prevAmount" fill={prevColor()}>
            {chartData.map((entry) => (
              <Cell key={entry.name} fill={prevColorAlpha()} stroke={prevColor()} strokeWidth={2} />
            ))}
          </Bar>
        )}
      </ComposedChart>
    </ChartWrapper>
  );
};
