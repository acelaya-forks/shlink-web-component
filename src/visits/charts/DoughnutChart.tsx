import { isDarkThemeEnabled, PRIMARY_DARK_COLOR, PRIMARY_LIGHT_COLOR } from '@shlinkio/shlink-frontend-kit';
import { formatNumber } from '@shlinkio/shlink-frontend-kit/tailwind';
import type { FC } from 'react';
import { useMemo } from 'react';
import { Cell, Pie, PieChart, Tooltip } from 'recharts';
import type { Stats } from '../types';
import { useChartDimensions } from './ChartDimensionsContext';
import { CHART_TOOLTIP_STYLES, chartColorForIndex, prevColor } from './constants';
import { DoughnutChartLegend } from './DoughnutChartLegend';

export type DoughnutChartProps = {
  stats: Stats;
  prevStats: Stats;
  showNumbersInLegend: boolean;
};

export type DoughnutChartEntry = {
  name: string;
  value: number;
  color: string;
};

const useChartData = (stats: Stats): DoughnutChartEntry[] => useMemo(
  () => Object.entries(stats).map(([name, value], i) => ({ name, value, color: chartColorForIndex(i) })),
  [stats],
);

export const DoughnutChart: FC<DoughnutChartProps> = ({ stats, prevStats, showNumbersInLegend }) => {
  const chartData = useChartData(stats);
  const prevChartData = useChartData(prevStats);
  const hasPrevCharts = prevChartData.length > 0;
  const borderColor = isDarkThemeEnabled() ? PRIMARY_DARK_COLOR : PRIMARY_LIGHT_COLOR;
  const { ChartWrapper, dimensions, wrapperDimensions } = useChartDimensions(300);

  return (
    <div className="tw:flex tw:flex-col tw:md:flex-row tw:md:items-center tw:gap-y-4">
      <div className="tw:md:flex-7">
        <div style={wrapperDimensions}>
          <ChartWrapper>
            <PieChart {...dimensions}>
              <Tooltip formatter={formatNumber} contentStyle={CHART_TOOLTIP_STYLES} itemStyle={{ color: 'white' }} />
              <Pie
                data={chartData}
                dataKey="value"
                nameKey="name"
                startAngle={360}
                endAngle={0}
                outerRadius="100%"
                innerRadius={hasPrevCharts ? '65%' : '50%'}
                animationBegin={0}
              >
                {chartData.map((entry, index) => (
                  <Cell key={`${entry.name}-${index}`} fill={entry.color} stroke={borderColor} />
                ))}
              </Pie>
              {hasPrevCharts && (
                <Pie
                  data={prevChartData}
                  dataKey="value"
                  nameKey="name"
                  startAngle={360}
                  endAngle={0}
                  outerRadius="55%"
                  innerRadius="20%"
                  animationBegin={0}
                  stroke={borderColor}
                  fill={prevColor()}
                />
              )}
            </PieChart>
          </ChartWrapper>
        </div>
      </div>
      <div className="tw:md:flex-5">
        <DoughnutChartLegend chartData={chartData} showNumbers={showNumbersInLegend} />
      </div>
    </div>
  );
};
