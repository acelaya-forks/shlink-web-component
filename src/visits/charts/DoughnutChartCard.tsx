import { useToggle } from '@shlinkio/shlink-frontend-kit';
import type { FC } from 'react';
import { SpaceBetweenContainer } from '../../common/SpaceBetweenContainer';
import { LabelledToggle } from '../../settings/components/fe-kit/LabelledToggle';
import { ChartCard } from './ChartCard';
import type { DoughnutChartProps } from './DoughnutChart';
import { DoughnutChart } from './DoughnutChart';

type DoughnutChartCardProps = Omit<DoughnutChartProps, 'showNumbersInLegend'> & {
  title: string;
};

export const DoughnutChartCard: FC<DoughnutChartCardProps> = ({ title, ...rest }) => {
  const { flag: showNumbersInLegend, toggle: toggleShowNumbersInLegend } = useToggle(false, true);

  return (
    <ChartCard
      title={(
        <SpaceBetweenContainer>
          {title}
          <LabelledToggle checked={showNumbersInLegend} onChange={toggleShowNumbersInLegend}>
            Show numbers
          </LabelledToggle>
        </SpaceBetweenContainer>
      )}
    >
      <DoughnutChart {...rest} showNumbersInLegend={showNumbersInLegend} />
    </ChartCard>
  );
};
