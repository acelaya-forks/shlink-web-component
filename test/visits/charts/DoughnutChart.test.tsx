import { screen } from '@testing-library/react';
import { ChartDimensionsProvider } from '../../../src/visits/charts/ChartDimensionsContext';
import { DoughnutChart } from '../../../src/visits/charts/DoughnutChart';
import { checkAccessibility } from '../../__helpers__/accessibility';
import { renderWithEvents } from '../../__helpers__/setUpTest';

describe('<DoughnutChart />', () => {
  const stats = { foo: 123, bar: 456 };
  const setUp = (prevStats = {}) => renderWithEvents(
    <ChartDimensionsProvider value={{ width: 800, height: 300 }}>
      <DoughnutChart stats={stats} prevStats={prevStats} showNumbersInLegend />
    </ChartDimensionsProvider>,
  );

  it('passes a11y checks', () => checkAccessibility(setUp()));

  it.each([
    [{}],
    [{ foo: 300, baz: 33 }],
    [{ ...stats, baz: 20 }],
  ])('renders Doughnut with expected props', (prevStats) => {
    const { container } = setUp(prevStats);
    expect(container).toMatchSnapshot();
  });

  it('renders expected legend', () => {
    setUp();

    expect(screen.getByText('foo')).toBeInTheDocument();
    expect(screen.getByText('bar')).toBeInTheDocument();
  });
});
