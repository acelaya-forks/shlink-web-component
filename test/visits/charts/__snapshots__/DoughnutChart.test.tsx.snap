// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<DoughnutChart /> > renders Doughnut with expected props 1`] = `
<div>
  <div
    class="tw:flex tw:flex-col tw:md:flex-row tw:md:items-center tw:gap-y-4"
  >
    <div
      class="tw:md:flex-7"
    >
      <div>
        <div
          class="recharts-wrapper"
          style="position: relative; cursor: default; width: 800px; height: 300px;"
        >
          <svg
            class="recharts-surface"
            cx="50%"
            cy="50%"
            height="300"
            style="width: 100%; height: 100%;"
            viewBox="0 0 800 300"
            width="800"
          >
            <title />
            <desc />
            <defs>
              <clippath
                id="recharts3-clip"
              >
                <rect
                  height="290"
                  width="790"
                  x="5"
                  y="5"
                />
              </clippath>
            </defs>
            <g
              class="recharts-layer recharts-pie"
              tabindex="0"
            >
              <g
                class="recharts-layer"
              >
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
              </g>
            </g>
          </svg>
          <div
            class="recharts-tooltip-wrapper recharts-tooltip-wrapper-right recharts-tooltip-wrapper-bottom"
            style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px; transform: translate(10px, 10px);"
            tabindex="-1"
          >
            <div
              class="recharts-default-tooltip"
              style="margin: 0px; padding: 10px; background: rgba(0, 0, 0, 0.9); border: none; white-space: nowrap; color: white; border-radius: 5px;"
            >
              <p
                class="recharts-tooltip-label"
                style="margin: 0px;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="tw:md:flex-5"
    >
      <ul
        class="tw:space-y-1"
      >
        <li
          class="tw:flex tw:items-center"
        >
          <div
            class="tw:inline-block tw:w-[20px] tw:h-[20px] tw:rounded-full tw:align-[-4px] tw:mr-1.5"
            data-testid="color-bullet-0"
            style="background-color: rgb(151, 187, 205);"
          />
          <small
            class="tw:truncate tw:grow"
            data-testid="name-0"
          >
            foo
            <b>
               (
              123
              )
            </b>
          </small>
        </li>
        <li
          class="tw:flex tw:items-center"
        >
          <div
            class="tw:inline-block tw:w-[20px] tw:h-[20px] tw:rounded-full tw:align-[-4px] tw:mr-1.5"
            data-testid="color-bullet-1"
            style="background-color: rgb(247, 70, 74);"
          />
          <small
            class="tw:truncate tw:grow"
            data-testid="name-1"
          >
            bar
            <b>
               (
              456
              )
            </b>
          </small>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<DoughnutChart /> > renders Doughnut with expected props 2`] = `
<div>
  <div
    class="tw:flex tw:flex-col tw:md:flex-row tw:md:items-center tw:gap-y-4"
  >
    <div
      class="tw:md:flex-7"
    >
      <div>
        <div
          class="recharts-wrapper"
          style="position: relative; cursor: default; width: 800px; height: 300px;"
        >
          <svg
            class="recharts-surface"
            cx="50%"
            cy="50%"
            height="300"
            style="width: 100%; height: 100%;"
            viewBox="0 0 800 300"
            width="800"
          >
            <title />
            <desc />
            <defs>
              <clippath
                id="recharts5-clip"
              >
                <rect
                  height="290"
                  width="790"
                  x="5"
                  y="5"
                />
              </clippath>
            </defs>
            <g
              class="recharts-layer recharts-pie"
              tabindex="0"
            >
              <g
                class="recharts-layer"
              >
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
              </g>
            </g>
            <g
              class="recharts-layer recharts-pie"
              tabindex="0"
            >
              <g
                class="recharts-layer"
              >
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
              </g>
            </g>
          </svg>
          <div
            class="recharts-tooltip-wrapper recharts-tooltip-wrapper-right recharts-tooltip-wrapper-bottom"
            style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px; transform: translate(10px, 10px);"
            tabindex="-1"
          >
            <div
              class="recharts-default-tooltip"
              style="margin: 0px; padding: 10px; background: rgba(0, 0, 0, 0.9); border: none; white-space: nowrap; color: white; border-radius: 5px;"
            >
              <p
                class="recharts-tooltip-label"
                style="margin: 0px;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="tw:md:flex-5"
    >
      <ul
        class="tw:space-y-1"
      >
        <li
          class="tw:flex tw:items-center"
        >
          <div
            class="tw:inline-block tw:w-[20px] tw:h-[20px] tw:rounded-full tw:align-[-4px] tw:mr-1.5"
            data-testid="color-bullet-0"
            style="background-color: rgb(151, 187, 205);"
          />
          <small
            class="tw:truncate tw:grow"
            data-testid="name-0"
          >
            foo
            <b>
               (
              123
              )
            </b>
          </small>
        </li>
        <li
          class="tw:flex tw:items-center"
        >
          <div
            class="tw:inline-block tw:w-[20px] tw:h-[20px] tw:rounded-full tw:align-[-4px] tw:mr-1.5"
            data-testid="color-bullet-1"
            style="background-color: rgb(247, 70, 74);"
          />
          <small
            class="tw:truncate tw:grow"
            data-testid="name-1"
          >
            bar
            <b>
               (
              456
              )
            </b>
          </small>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<DoughnutChart /> > renders Doughnut with expected props 3`] = `
<div>
  <div
    class="tw:flex tw:flex-col tw:md:flex-row tw:md:items-center tw:gap-y-4"
  >
    <div
      class="tw:md:flex-7"
    >
      <div>
        <div
          class="recharts-wrapper"
          style="position: relative; cursor: default; width: 800px; height: 300px;"
        >
          <svg
            class="recharts-surface"
            cx="50%"
            cy="50%"
            height="300"
            style="width: 100%; height: 100%;"
            viewBox="0 0 800 300"
            width="800"
          >
            <title />
            <desc />
            <defs>
              <clippath
                id="recharts8-clip"
              >
                <rect
                  height="290"
                  width="790"
                  x="5"
                  y="5"
                />
              </clippath>
            </defs>
            <g
              class="recharts-layer recharts-pie"
              tabindex="0"
            >
              <g
                class="recharts-layer"
              >
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
              </g>
            </g>
            <g
              class="recharts-layer recharts-pie"
              tabindex="0"
            >
              <g
                class="recharts-layer"
              >
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
              </g>
            </g>
          </svg>
          <div
            class="recharts-tooltip-wrapper recharts-tooltip-wrapper-right recharts-tooltip-wrapper-bottom"
            style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px; transform: translate(10px, 10px);"
            tabindex="-1"
          >
            <div
              class="recharts-default-tooltip"
              style="margin: 0px; padding: 10px; background: rgba(0, 0, 0, 0.9); border: none; white-space: nowrap; color: white; border-radius: 5px;"
            >
              <p
                class="recharts-tooltip-label"
                style="margin: 0px;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="tw:md:flex-5"
    >
      <ul
        class="tw:space-y-1"
      >
        <li
          class="tw:flex tw:items-center"
        >
          <div
            class="tw:inline-block tw:w-[20px] tw:h-[20px] tw:rounded-full tw:align-[-4px] tw:mr-1.5"
            data-testid="color-bullet-0"
            style="background-color: rgb(151, 187, 205);"
          />
          <small
            class="tw:truncate tw:grow"
            data-testid="name-0"
          >
            foo
            <b>
               (
              123
              )
            </b>
          </small>
        </li>
        <li
          class="tw:flex tw:items-center"
        >
          <div
            class="tw:inline-block tw:w-[20px] tw:h-[20px] tw:rounded-full tw:align-[-4px] tw:mr-1.5"
            data-testid="color-bullet-1"
            style="background-color: rgb(247, 70, 74);"
          />
          <small
            class="tw:truncate tw:grow"
            data-testid="name-1"
          >
            bar
            <b>
               (
              456
              )
            </b>
          </small>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
